File: frontend/src/lib/api/knowledge.ts
import { apiDelete, apiGet, apiPost, apiPostForm } from "./core";

export type RAGEngineType = "ragflow" | "pgvector";
export type KnowledgeBaseStatus = "active" | "archived";
export type KnowledgeBaseVisibility = "private" | "department" | "tenant";
export type KnowledgeDocumentStatus = "pending_upload" | "uploaded" | "ingesting" | "indexed" | "failed" | "deleted";
export type KnowledgeDocumentSource = "api_upload" | "channel_attachment" | "internal_import";

export interface RetrievalConfig {
  top_k: number;
  score_threshold: number | null;
  rerank_enabled: boolean;
  citation_required: boolean;
  metadata_filters: Record<string, unknown>;
}

export interface KnowledgeBaseCreateRequest {
  name: string;
  description?: string | null;
  visibility: KnowledgeBaseVisibility;
  department_ids?: string[];
  rag_engine: RAGEngineType;
  embedding_model_key?: string | null;
  retrieval_config?: RetrievalConfig;
  tags?: string[];
  metadata?: Record<string, unknown>;
}

export interface KnowledgeBaseResponse {
  id: string;
  tenant_id: string;
  name: string;
  description: string | null;
  visibility: KnowledgeBaseVisibility;
  department_ids: string[];
  rag_engine: RAGEngineType;
  embedding_model_key: string | null;
  retrieval_config: RetrievalConfig;
  status: KnowledgeBaseStatus;
  document_count: number;
  tags: string[];
  metadata: Record<string, unknown>;
  created_at: string;
  updated_at: string;
}

export interface KnowledgeBaseListResponse {
  bases: KnowledgeBaseResponse[];
}

export interface WorkbenchKnowledgeBaseResponse {
  id: string;
  name: string;
  description: string | null;
  visibility: KnowledgeBaseVisibility;
  department_ids: string[];
  status: KnowledgeBaseStatus;
  document_count: number;
  tags: string[];
  updated_at: string;
}

export interface WorkbenchKnowledgeBaseListResponse {
  bases: WorkbenchKnowledgeBaseResponse[];
}

export interface KnowledgeGovernanceTargetItem {
  id: string;
  label: string;
  description: string | null;
  metadata: Record<string, unknown>;
}

export interface KnowledgeGovernanceTargetsResponse {
  departments: KnowledgeGovernanceTargetItem[];
}

export interface KnowledgeDocumentResponse {
  id: string;
  knowledge_base_id: string;
  tenant_id: string;
  filename: string;
  content_type: string | null;
  size_bytes: number | null;
  checksum_sha256: string | null;
  source: KnowledgeDocumentSource;
  status: KnowledgeDocumentStatus;
  storage_bucket: string;
  storage_object_key: string;
  rag_document_id: string | null;
  chunk_count: number;
  error_message: string | null;
  metadata: Record<string, unknown>;
  created_at: string;
  updated_at: string;
}

export interface KnowledgeDocumentListResponse {
  documents: KnowledgeDocumentResponse[];
}

export interface WorkbenchKnowledgeDocumentResponse {
  id: string;
  knowledge_base_id: string;
  filename: string;
  content_type: string | null;
  size_bytes: number | null;
  source: KnowledgeDocumentSource;
  status: KnowledgeDocumentStatus;
  chunk_count: number;
  updated_at: string;
}

export interface WorkbenchKnowledgeDocumentListResponse {
  documents: WorkbenchKnowledgeDocumentResponse[];
}

export interface KnowledgeDeleteResponse {
  id: string;
  deleted: boolean;
  message: string;
  diagnostics: Record<string, unknown>;
}

export interface RetrievalTestRequest {
  query: string;
  top_k: number;
  score_threshold?: number | null;
  filters?: Record<string, unknown>;
  include_raw_chunks?: boolean;
  rerank?: boolean;
}

export interface RetrievalSourceResponse {
  chunk_id: string;
  document_id: string | null;
  source_name: string | null;
  score: number | null;
  text: string;
  metadata: Record<string, unknown>;
}

export interface RetrievalTestResponse {
  knowledge_base_id: string;
  query: string;
  engine: RAGEngineType;
  elapsed_ms: number;
  results: RetrievalSourceResponse[];
  diagnostics: Record<string, unknown>;
  checked_at: string;
}

export interface DocumentUploadCompleteRequest {
  etag?: string | null;
  size_bytes?: number | null;
  checksum_sha256?: string | null;
  auto_ingest: boolean;
  metadata?: Record<string, unknown>;
}

export interface DocumentUploadCompleteResponse {
  document: KnowledgeDocumentResponse;
  auto_ingest: boolean;
  ingest_status: KnowledgeDocumentStatus | null;
  message: string;
  diagnostics: Record<string, unknown>;
}

export const knowledgeApi = {
  getKnowledgeGovernanceTargets: () =>
    apiGet<KnowledgeGovernanceTargetsResponse>("/api/v1/knowledge/governance-targets"),
  getKnowledgeBases: () => apiGet<KnowledgeBaseListResponse>("/api/v1/knowledge/bases"),
  getWorkbenchKnowledgeBases: () => apiGet<WorkbenchKnowledgeBaseListResponse>("/api/v1/knowledge/workbench/bases"),
  createKnowledgeBase: (payload: KnowledgeBaseCreateRequest) =>
    apiPost<KnowledgeBaseResponse, KnowledgeBaseCreateRequest>("/api/v1/knowledge/bases", payload),
  deleteKnowledgeBase: (baseId: string) => apiDelete<KnowledgeDeleteResponse>(`/api/v1/knowledge/bases/${baseId}`),
  getKnowledgeDocuments: (baseId: string) =>
    apiGet<KnowledgeDocumentListResponse>(`/api/v1/knowledge/bases/${baseId}/documents`),
  getWorkbenchKnowledgeDocuments: (baseId: string) =>
    apiGet<WorkbenchKnowledgeDocumentListResponse>(`/api/v1/knowledge/workbench/bases/${baseId}/documents`),
  deleteKnowledgeDocument: (baseId: string, documentId: string) =>
    apiDelete<KnowledgeDeleteResponse>(`/api/v1/knowledge/bases/${baseId}/documents/${documentId}`),
  reingestKnowledgeDocument: (baseId: string, documentId: string) =>
    apiPost<DocumentUploadCompleteResponse, Record<string, never>>(
      `/api/v1/knowledge/bases/${baseId}/documents/${documentId}/reingest`,
      {},
    ),
  uploadKnowledgeDocument: (baseId: string, file: File, autoIngest = true) => {
    const formData = new FormData();
    formData.append("file", file);
    formData.append("auto_ingest", String(autoIngest));
    return apiPostForm<DocumentUploadCompleteResponse>(`/api/v1/knowledge/bases/${baseId}/documents/upload`, formData);
  },
  completeKnowledgeDocumentUpload: (baseId: string, documentId: string, payload: DocumentUploadCompleteRequest) =>
    apiPost<DocumentUploadCompleteResponse, DocumentUploadCompleteRequest>(
      `/api/v1/knowledge/bases/${baseId}/documents/${documentId}/complete-upload`,
      payload,
    ),
  runRetrievalTest: (baseId: string, payload: RetrievalTestRequest) =>
    apiPost<RetrievalTestResponse, RetrievalTestRequest>(`/api/v1/knowledge/bases/${baseId}/retrieval-test`, payload),
};

File: backend/app/schemas/knowledge.py
from datetime import datetime, timezone
from enum import StrEnum
from typing import Any
from uuid import UUID

from pydantic import BaseModel, Field, field_validator, model_validator

from app.rag.schemas import RAGEngineType


class KnowledgeBaseStatus(StrEnum):
    ACTIVE = "active"
    ARCHIVED = "archived"


class KnowledgeBaseVisibility(StrEnum):
    PRIVATE = "private"
    DEPARTMENT = "department"
    TENANT = "tenant"


class KnowledgeDocumentStatus(StrEnum):
    PENDING_UPLOAD = "pending_upload"
    UPLOADED = "uploaded"
    INGESTING = "ingesting"
    INDEXED = "indexed"
    FAILED = "failed"
    DELETED = "deleted"


class KnowledgeDocumentSource(StrEnum):
    API_UPLOAD = "api_upload"
    CHANNEL_ATTACHMENT = "channel_attachment"
    INTERNAL_IMPORT = "internal_import"


class RetrievalConfig(BaseModel):
    top_k: int = Field(default=5, ge=1, le=50)
    score_threshold: float | None = Field(default=None, ge=0, le=1)
    rerank_enabled: bool = False
    citation_required: bool = True
    metadata_filters: dict[str, Any] = Field(default_factory=dict)


class KnowledgeBaseCreateRequest(BaseModel):
    name: str = Field(min_length=1, max_length=120)
    description: str | None = Field(default=None, max_length=1000)
    visibility: KnowledgeBaseVisibility = KnowledgeBaseVisibility.TENANT
    department_ids: list[UUID] = Field(default_factory=list)
    rag_engine: RAGEngineType = RAGEngineType.RAGFLOW
    embedding_model_key: str | None = Field(default=None, max_length=120)
    retrieval_config: RetrievalConfig = Field(default_factory=RetrievalConfig)
    tags: list[str] = Field(default_factory=list, max_length=20)
    metadata: dict[str, Any] = Field(default_factory=dict)

    @model_validator(mode="after")
    def validate_visibility_scope(self) -> "KnowledgeBaseCreateRequest":
        if self.visibility == KnowledgeBaseVisibility.DEPARTMENT and not self.department_ids:
            raise ValueError("department_ids is required when visibility is department.")
        if self.visibility != KnowledgeBaseVisibility.DEPARTMENT:
            self.department_ids = []
        return self


class KnowledgeBaseResponse(BaseModel):
    id: UUID
    tenant_id: UUID
    name: str
    description: str | None
    visibility: KnowledgeBaseVisibility
    department_ids: list[UUID]
    rag_engine: RAGEngineType
    embedding_model_key: str | None
    retrieval_config: RetrievalConfig
    status: KnowledgeBaseStatus
    document_count: int = Field(ge=0)
    tags: list[str]
    metadata: dict[str, Any]
    created_at: datetime
    updated_at: datetime


class KnowledgeBaseListResponse(BaseModel):
    bases: list[KnowledgeBaseResponse]


class WorkbenchKnowledgeBaseResponse(BaseModel):
    id: UUID
    name: str
    description: str | None
    visibility: KnowledgeBaseVisibility
    department_ids: list[UUID]
    status: KnowledgeBaseStatus
    document_count: int = Field(ge=0)
    tags: list[str]
    updated_at: datetime


class WorkbenchKnowledgeBaseListResponse(BaseModel):
    bases: list[WorkbenchKnowledgeBaseResponse]


class KnowledgeGovernanceTargetItem(BaseModel):
    id: UUID
    label: str
    description: str | None = None
    metadata: dict[str, Any] = Field(default_factory=dict)


class KnowledgeGovernanceTargetsResponse(BaseModel):
    departments: list[KnowledgeGovernanceTargetItem] = Field(default_factory=list)


class DocumentUploadPrepareRequest(BaseModel):
    filename: str = Field(min_length=1, max_length=255)
    content_type: str | None = Field(default=None, max_length=160)
    size_bytes: int | None = Field(default=None, ge=0, le=1024 * 1024 * 1024)
    checksum_sha256: str | None = Field(default=None, min_length=64, max_length=64)
    source: KnowledgeDocumentSource = KnowledgeDocumentSource.API_UPLOAD
    parser_config: dict[str, Any] = Field(default_factory=dict)
    metadata: dict[str, Any] = Field(default_factory=dict)

    @field_validator("filename")
    @classmethod
    def filename_must_not_be_path(cls, value: str) -> str:
        normalized = value.strip()
        if not normalized or "/" in normalized or "\\" in normalized:
            raise ValueError("filename must be a plain file name, not a path")
        return normalized


class KnowledgeDocumentResponse(BaseModel):
    id: UUID
    knowledge_base_id: UUID
    tenant_id: UUID
    filename: str
    content_type: str | None
    size_bytes: int | None
    checksum_sha256: str | None
    source: KnowledgeDocumentSource
    status: KnowledgeDocumentStatus
    storage_bucket: str
    storage_object_key: str
    rag_document_id: str | None
    chunk_count: int = Field(ge=0)
    error_message: str | None = None
    metadata: dict[str, Any] = Field(default_factory=dict)
    created_at: datetime
    updated_at: datetime


class KnowledgeDocumentListResponse(BaseModel):
    documents: list[KnowledgeDocumentResponse]


class WorkbenchKnowledgeDocumentResponse(BaseModel):
    id: UUID
    knowledge_base_id: UUID
    filename: str
    content_type: str | None
    size_bytes: int | None
    source: KnowledgeDocumentSource
    status: KnowledgeDocumentStatus
    chunk_count: int = Field(ge=0)
    updated_at: datetime


class WorkbenchKnowledgeDocumentListResponse(BaseModel):
    documents: list[WorkbenchKnowledgeDocumentResponse]


class KnowledgeDeleteResponse(BaseModel):
    id: UUID
    deleted: bool
    message: str
    diagnostics: dict[str, Any] = Field(default_factory=dict)


class StorageTargetResponse(BaseModel):
    bucket: str
    object_key: str
    content_type: str | None = None
    size_bytes: int | None = Field(default=None, ge=0)
    checksum_sha256: str | None = None
    upload_url: str | None = None
    expires_in_seconds: int | None = None
    headers: dict[str, str] = Field(default_factory=dict)
    placeholder: bool = True


class DocumentIngestPlanResponse(BaseModel):
    rag_engine: RAGEngineType
    vector_store: str = "pgvector"
    parser_config: dict[str, Any] = Field(default_factory=dict)
    auto_ingest: bool = False
    message: str


class DocumentUploadPrepareResponse(BaseModel):
    document: KnowledgeDocumentResponse
    upload_session_id: str
    storage: StorageTargetResponse
    ingest_plan: DocumentIngestPlanResponse


class DocumentUploadCompleteRequest(BaseModel):
    etag: str | None = Field(default=None, max_length=256)
    size_bytes: int | None = Field(default=None, ge=0, le=1024 * 1024 * 1024)
    checksum_sha256: str | None = Field(default=None, min_length=64, max_length=64)
    auto_ingest: bool = True
    metadata: dict[str, Any] = Field(default_factory=dict)


class DocumentUploadCompleteResponse(BaseModel):
    document: KnowledgeDocumentResponse
    auto_ingest: bool
    ingest_status: KnowledgeDocumentStatus | None = None
    message: str
    diagnostics: dict[str, Any] = Field(default_factory=dict)


class RetrievalTestRequest(BaseModel):
    query: str = Field(min_length=1, max_length=4000)
    top_k: int = Field(default=5, ge=1, le=50)
    score_threshold: float | None = Field(default=None, ge=0, le=1)
    filters: dict[str, Any] = Field(default_factory=dict)
    include_raw_chunks: bool = False
    rerank: bool = False


class RetrievalSourceResponse(BaseModel):
    chunk_id: str
    document_id: UUID | None = None
    source_name: str | None = None
    score: float | None = None
    text: str
    metadata: dict[str, Any] = Field(default_factory=dict)


class RetrievalTestResponse(BaseModel):
    knowledge_base_id: UUID
    query: str
    engine: RAGEngineType
    elapsed_ms: int = Field(ge=0)
    results: list[RetrievalSourceResponse]
    diagnostics: dict[str, Any] = Field(default_factory=dict)
    checked_at: datetime = Field(default_factory=lambda: datetime.now(timezone.utc))
